### PR TITLE
Add GA4 mock functions.

### PR DESCRIPTION
**Pull request by asvinb**
## Summary

Addresses issue:

- #6174 

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+	);
+
+	// We set seed for every data mock to make sure that the same arguments get the same report data.
+	// It means that everyone will have the same report data and will see the same widgets in the storybook.
+	// This approach gives us additional flexibility to control randomness on a per widget basis.
+	if ( ! Number.isNaN( argsHash ) ) {
+		faker.seed( argsHash );
+	}
+
+	const data = {
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:219]: Hi @asvinb, unfortunately a bit of detail was missed in the IB. We need to ensure the response shape matches the new [response schema](https://developers.google.com/analytics/devguides/reporting/data/v1/rest/v1beta/RunReportResponse). Therefore a bit more work is needed here.

As a concrete reference, using the [API Explorer](https://developers.google.com/analytics/devguides/reporting/data/v1/rest/v1beta/properties/runReport) I was able to generate an example report:

<details>
<summary>Example report</summary>

```json
{
  "dimensionHeaders": [
    {
      "name": "date"
    }
  ],
  "metricHeaders": [
    {
      "name": "sessions",
      "type": "TYPE_INTEGER"
    },
    {
      "name": "bounceRate",
      "type": "TYPE_FLOAT"
    }
  ],
  "rows": [
    {
      "dimensionValues": [
        {
          "value": "20230101"
        }
      ],
      "metricValues": [
        {
          "value": "55"
        },
        {
          "value": "0.61818181818181817"
        }
      ]
    },
    {
      "dimensionValues": [
        {
          "value": "20230102"
        }
      ],
      "metricValues": [
        {
          "value": "7"
        },
        {
          "value": "0.7142857142857143"
        }
      ]
    }
  ],
  "totals": [
    {
      "dimensionValues": [
        {
          "value": "RESERVED_TOTAL"
        }
      ],
      "metricValues": [
        {
          "value": "62"
        },
        {
          "value": "0.62903225806451613"
        }
      ]
    }
  ],
  "maximums": [
    {
      "dimensionValues": [
        {
          "value": "RESERVED_MAX"
        }
      ],
      "metricValues": [
        {
          "value": "55"
        },
        {
          "value": "0.7142857142857143"
        }
      ]
    }
  ],
  "minimums": [
    {
      "dimensionValues": [
        {
          "value": "RESERVED_MIN"
        }
      ],
      "metricValues": [
        {
          "value": "7"
        },
        {
          "value": "0.61818181818181817"
        }
      ]
    }
  ],
  "rowCount": 2,
  "metadata": {
    "currencyCode": "USD",
    "timeZone": "America/Los_Angeles"
  },
  "kind": "analyticsData#runReport"
}
```

</details>

<details>
<summary>Query for the above report</summary>

```json
{
  "dimensions": [
    {
      "name": "date"
    }
  ],
  "metrics": [
    {
      "name": "sessions"
    },
    {
      "name": "bounceRate"
    }
  ],
  "metricAggregations": [
    "TOTAL",
    "MINIMUM",
    "MAXIMUM"
  ],
  "dateRanges": [
    {
      "startDate": "2023-01-01",
      "endDate": "2023-01-03"
    }
  ]
}
```

</details>

**Review by techanvil — CHANGES_REQUESTED** [review]
Hi @asvinb, this is looking good, but as mentioned a bit of detail was missed and this needs a bit more work to ensure the response shape matches the API response.

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+import { MODULES_ANALYTICS_4 } from '../datastore/constants';
+import { isValidDateString } from '../../../util';
+import { stringToDate } from '../../../util/date-range/string-to-date';
+
+const ANALYTICS_4_METRIC_TYPES = {
+	totalUsers: 'INTEGER',
+	newUsers: 'INTEGER',
+	sessions: 'INTEGER',
+	screenPageViews: 'INTEGER',
+	bounceRate: 'PERCENT',
```
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: 👍

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+	averageSessionDuration: 'TIME',
+};
```
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js:45]: Thanks @techanvil . Updated!

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+
+const ANALYTICS_4_DIMENSION_OPTIONS = {
+	sessionDefaultChannelGrouping: [
+		'Organic Search',
+		'Referral',
+		'Direct',
+		'(other)',
```
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: Updated @techanvil

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+		'United States',
+		'United Kingdom',
+		'India',
+		'(not set)',
+		'France',
+		'Ukraine',
+		'Italy',
+		'Mexico',
+	],
+	deviceCategory: [ 'desktop', 'tablet', 'mobile' ],
```
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: 👍

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 1 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js:230]: Thanks @techanvil . I knew something was kinda missing 😅  I've updated the PR, can you take another look?

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+		maximums: [],
+		metadata: {
+			currencyCode: 'USD',
+			timeZone: 'America/Los_Angeles',
+		},
+		kind: 'analyticsData#runReport',
+	};
+
+	const { compareStartDate, compareEndDate } = args;
+	const metricValuesCount = compareStartDate && compareEndDate ? 2 : 1;
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:244]: The usage of `metricValuesCount` and related logic needs to be changed (and it should probably be renamed), due to the difference in response shape for GA4 whereby there are additional rows for each date range, rather than additional entries per-row. Also, and on a related note, the aggregate properties have a different structure. Please see the example GA4 response in this https://github.com/google/site-kit-wp/issues/6172#issuecomment-1375638193 for details.

**Review by techanvil — CHANGES_REQUESTED** [review]
Hi @asvinb, thanks for making the changes here. However, since implementing #6172 I have, identified some more changes that need to be made. Please see the related comments.

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+import { isValidDateString } from '../../../util';
+import { stringToDate } from '../../../util/date-range/string-to-date';
+
+const ANALYTICS_4_METRIC_TYPES = {
+	totalUsers: 'INTEGER',
+	newUsers: 'INTEGER',
+	sessions: 'INTEGER',
+	screenPageViews: 'INTEGER',
+	bounceRate: 'PERCENT',
+	averageSessionDuration: 'TIME',
```
[the remark this answers is not in the document]
Comment (techanvil) [reply assets/js/modules/analytics-4/utils/data-mock.js:44]: The PRD has been updated so `averageSessionDuration` is confirmed :+1:

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 6 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: Thanks @techanvil . This is ready for you to review again.

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+import { stringToDate } from '../../../util/date-range/string-to-date';
+
+const ANALYTICS_4_METRIC_TYPES = {
+	totalUsers: 'INTEGER',
+	newUsers: 'INTEGER',
+	sessions: 'INTEGER',
+	conversions: 'INTEGER',
+	screenPageViews: 'INTEGER',
+	engagedSessions: 'INTEGER',
+	averageSessionDuration: 'TIME',
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:44]: Hey @asvinb, it turns out we need to update the type names here seeing as they end up in the response, they should be as per this screen grab:

image

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+	faker.seed( originalSeedValue );
+
+	return [
+		{
+			dimensionHeaders:
+				args?.dimensions?.map( ( dimension ) => ( {
+					name: dimension,
+				} ) ) || null,
+			metricHeaders: validMetrics.map( ( metric ) => ( {
+				name: metric?.alias || metric?.expression || metric.toString(),
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:448]: We should no longer have `alias`, but do have `name` instead. We should however never have a metric with an `expression` but no `name` (as described in https://github.com/google/site-kit-wp/issues/6172#issuecomment-1375638193).

So I think this could be changed as follows:

```suggestion
				name: metric?.name || metric.toString(),
```

File: assets/js/modules/analytics-4/utils/data-mock.test.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+					expression: 'sessions',
+				},
+				{
+					expression: 'newUsers',
+				},
+			],
+			dimensions: [ 'date' ],
+		} );
+
+		expect( report[ 0 ].rows ).toHaveLength( 28 );
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.test.js:61]: I should have called this out the first time really, it would be useful to have some JSON test fixtures for the report output in these tests, which we could do deep equality checks against.

This would be a much more robust test, and also serve to illustrate the output. Also, when I log out `report` in these tests, it has empty arrays for `metricValues` which is a bit odd.

I'd also suggest providing a wider range of report params so we can get more comprehensive test coverage.

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+		'India',
+		'(not set)',
+		'France',
+		'Ukraine',
+		'Italy',
+		'Mexico',
+	],
+	deviceCategory: [ 'Desktop', 'Tablet', 'Mobile' ],
+	pageTitle: ( i ) => ( i <= 12 ? `Test Post ${ i }` : false ),
+	pagePathPlusQueryString: ( i ) =>
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:72]: Seeing as we've switched from `pagePathPlusQueryString` to `pagePath` in the `GET:report` endpoint due to compatibility issues, we should probably also update this to `pagePath`:
```suggestion
	pagePath: ( i ) =>
```

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+					{
+						value: faker.datatype
+							.float( { min: 0, max: 10000 } )
+							.toString(),
+					},
+				]; // $10k max.
+		}
+	} );
+
+	return [];
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:147]: Ah, this function is now hardwired to return an empty array - this is why `metricValues` is empty as per my other comment.

**Review by techanvil — CHANGES_REQUESTED** [review]
Hey @asvinb, thanks for updating this. I've spotted a couple of new things and some I missed the first time around. Please take another look.

File: assets/js/modules/analytics-4/utils/data-mock.test.js (hunk 10 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.test.js]: @techanvil I updated the tests, can you take a look and let me know what you think?

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 9 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: 👍

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 8 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: 👍

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 11 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: 👍

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 12 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: 👍

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 13]
```diff
@@ -376,63 +386,143 @@ export function getAnalytics4MockResponse( args ) {
 		),
 	];
 
+	const metricAggregations = castArray( args.metricAggregations );
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:389]: Metric aggregations are currently unconditionally requested for all reports, so we should always include them here rather than making it an argument.

https://github.com/google/site-kit-wp/blob/4cb5777fae8d0d724629872917f13dbecf0424e4/includes/Modules/Analytics_4.php#L1083-L1090

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 14]
```diff
@@ -0,0 +1,558 @@
...
+ *
+ * @param {Array.<Object>} validMetrics Metric list.
+ * @return {Array.<Object>} Array of metric values.
+ */
+function generateMetricValues( validMetrics ) {
+	const values = [];
+
+	validMetrics.forEach( ( validMetric ) => {
+		switch ( getMetricType( validMetric ) ) {
+			case 'INTEGER':
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:112]: I think it would be preferable to include the full type here rather than prefixing `TYPE_` further on. That makes it a bit clearer when cross referencing the report output with this code. Also it improves searchability, so this file will appear in the search results when searching for `TYPE_INTEGER` etc.

```suggestion
			case 'TYPE_INTEGER':
```

Obviously, this applies to the other types below.

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 15]
```diff
@@ -0,0 +1,558 @@
...
+
+	return [
+		{
+			dimensionHeaders:
+				args?.dimensions?.map( ( dimension ) => ( {
+					name: dimension,
+				} ) ) || null,
+			metricHeaders: validMetrics.map( ( metric ) => ( {
+				name: metric?.name || metric.toString(),
+				type: `TYPE_${ getMetricType( metric ) }`,
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:539]: As mentioned above, it would be preferable to avoid this string concatenation and instead use the full string values throughout.

```suggestion
				type: `TYPE_${ getMetricType( metric ) }`,
```

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 16]
```diff
@@ -0,0 +1,558 @@
...
+			currencyCode: 'USD',
+			timeZone: 'America/Los_Angeles',
+		},
+		kind: 'analyticsData#runReport',
+	};
+
+	const { compareStartDate, compareEndDate } = args;
+	const hasDateRange = compareStartDate && compareEndDate;
+
+	const validMetrics = ( args.metrics || [] ).filter(
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:267]: I think we should pass the full type names in `args.metrics`, as discussed elsewhere. So `args.metrics` should contain `TYPE_INTEGER`, `TYPE_STRING` etc. I think this would be preferable as it keeps things closer to the real report data, even if it means the interface is slightly different to the UA mock utility.

File: assets/js/modules/analytics-4/datastore/__fixtures__/report.json [hunk 17]
```diff
@@ -0,0 +1,1961 @@
+{
```
Comment (techanvil) [inline assets/js/modules/analytics-4/datastore/__fixtures__/report.json:1]: This report fixture is not being used anywhere, I would have expected to see it used in a test to compare the mock output...

It could also be worth adding and testing against a fixture which only uses a single date range, and also including a wider variety of data types...

**Review by techanvil — CHANGES_REQUESTED** [review]
Hi @asvinb, thanks for making changes, there are still a few items to address, please take a look.

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 13 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: 👍

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 14 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: Sounds good. Updated!

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 15 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js]: @techanvil Here we are dynamically getting the type, so we still need to call the `getMetricType` function but without the `TYPE_` prefix.

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 16 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.js:249]: @techanvil Can you be more explicit here please? 😅  I would prefer to have the metrics to be in the same format as the API, i.e no need to have the types in there.

File: assets/js/modules/analytics-4/datastore/__fixtures__/report.json (hunk 17 above)
Comment (asvinb) [reply assets/js/modules/analytics-4/datastore/__fixtures__/report.json:1]: @techanvil I updated the tests but maybe I'll need some help after you review about possible tests to write. Right now I'am mostly checking the format of `dimensionHeaders`, `metricsHeaders`, etc

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 16 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/utils/data-mock.js:249]: Hi @asvinb, my apologies, my comment was complete gibberish, I had misread the code. Please ignore!

File: assets/js/modules/analytics-4/utils/data-mock.js (hunk 15 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/utils/data-mock.js]: That sounds fine to me, my point was really just a reminder to update following the comment above.

File: assets/js/modules/analytics-4/utils/data-mock.test.js [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+		);
+		expect( report[ 0 ].rows[ 0 ].dimensionValues ).toHaveLength(
+			reportResponse.rows[ 0 ].dimensionValues.length
+		);
+		expect( report[ 0 ].rows[ 0 ].metricValues ).toHaveLength(
+			reportResponse.rows[ 0 ].metricValues.length
+		);
+	} );
+
+	it( 'checks if the report structure matches what we get from the response when there is a date range', () => {
```
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.test.js]: 👍

File: assets/js/modules/analytics-4/utils/data-mock.test.js [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+
+	it( 'checks if the report structure matches what we get from the response', () => {
+		const report = getAnalytics4MockResponse( {
+			startDate: '2022-11-01',
+			endDate: '2022-12-31',
+			metrics: [ 'totalUsers', 'averageSessionDuration', 'sessions' ],
+			dimensions: [ 'date', 'country' ],
+		} );
+
+		expect( report[ 0 ].dimensionHeaders ).toMatchObject(
```
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.test.js:158]: @techanvil I don't think we can use `toEqual` to check between the response and the data generated since the data generated is dynamic and will never match what we have in the JSON file. That's why I compared the length of some elements.
I've also added test to verify that the number of date_range_0 rows matches the number of date_range_1 rows

[the remark this answers is not in the document]
Comment (techanvil) [reply assets/js/modules/analytics-4/utils/data-mock.test.js:158]: Hey @asvinb, actually, as described below the RNG utility has an explicit seed set so the generated values are the same for each combination of report args. I gave it a quick test by snapshotting the report and it does work. So, we should be able to use `toEqual` here.

https://github.com/google/site-kit-wp/blob/0f06107080a26ad10d8b855bdcc8261472841331/assets/js/modules/analytics-4/utils/data-mock.js#L226-L228

**Review by techanvil — CHANGES_REQUESTED** [review]
Thanks for updating @asvinb. Please see my comment about the randomizer / `toEqual` usage.

File: assets/js/modules/analytics-4/utils/data-mock.test.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/analytics-4/utils/data-mock.test.js:160]: @techanvil I've updated the PR, can you take another look please?
Thanks!

File: assets/js/modules/analytics-4/datastore/__fixtures__/report-mutiple-date-ranges.json [hunk 20]
```diff
@@ -0,0 +1,1961 @@
+{
```
Comment (techanvil) [inline assets/js/modules/analytics-4/datastore/__fixtures__/report-mutiple-date-ranges.json:1]: This file can be removed...

File: assets/js/modules/analytics-4/datastore/__fixtures__/report.json [hunk 21]
```diff
@@ -0,0 +1,878 @@
...
```
Comment (techanvil) [inline assets/js/modules/analytics-4/datastore/__fixtures__/report.json:1]: This file can be removed...

File: assets/js/modules/analytics-4/utils/__fixtures__/mocked-report-multiple-date-ranges.json [hunk 22]
```diff
@@ -0,0 +1 @@
...]},{"dimensionValues":[{"value":"20201128"},{"value":"date_range_1"}],"metricValues":[{"value":"53"},{"value":"7"}]},{"dimensionValues":[{"value":"20201129"},{"value":"date_range_0"}],"metricValues":[{"value":"83"},{"value":"8"}]},{"dimensionValues":[{"value":"20201129"},{"value":"date_range_1"}],"metricValues":[{"value":"72"},{"value":"34"}]}],"totals":[{"dimensionValues":[{"value":"RESERVED_TOTAL"},{"value":"date_range_0"}],"metricValues":[{"value":"72"},{"value":"34"}]},{"dimensionValues":[{"value":"RESERVED_TOTAL"},{"value":"date_range_1"}],"metricValues":[{"value":"72"},{"value":"34"}]}],"minimums":[{"dimensionValues":[{"value":"RESERVED_MIN"},{"value":"date_range_0"}],"metricValues":[{"value":"55"},{"value":"8"}]},{"dimensionValues":[{"value":"RESERVED_MIN"},{"value":"date_range_1"}],"metricValues":[{"value":"55"},{"value":"8"}]}],"maximums":[{"dimensionValues":[{"value":"RESERVED_MAX"},{"value":"date_range_0"}],"metricValues":[{"value":"72"},{"value":"34"}]},{"dimensionValues":[{"value":"RESERVED_MAX"},{"value":"date_range_1"}],"metricValues":[{"value":"72"},{"value":"34"}]}],"metadata":{"currencyCode":"USD","timeZone":"America/Los_Angeles"},"kind":"analyticsData#runReport"}
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/__fixtures__/mocked-report-multiple-date-ranges.json:1]: Please can you pretty-print this file?

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 23]
```diff
@@ -0,0 +1,532 @@
...
+					  ]
+					: []
+			);
+		} );
+
+	// Set the original seed value for the faker.
+	faker.seed( originalSeedValue );
+
+	return [
+		{
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:506]: Sorry for not spotting this earlier, but this needs to be changed to only return a single object, rather than array with a single report object inside. There's a bit more info [here](https://github.com/google/site-kit-wp/issues/6172#issuecomment-1375638193) under "Single report".

This will obviously have a knock-on effect to the tests.

```suggestion
	return {
```

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 24]
```diff
@@ -0,0 +1,532 @@
...
+				args?.dimensions?.map( ( dimension ) => ( {
+					name: dimension,
+				} ) ) || null,
+			metricHeaders: validMetrics.map( ( metric ) => ( {
+				name: metric?.name || metric.toString(),
+				type: getMetricType( metric ),
+			} ) ),
+			...data,
+		},
+	];
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:517]: As above:

```suggestion
	};
```

**Review by techanvil — CHANGES_REQUESTED** [review]
Hi @asvinb, this needs a couple of tweaks to finish up, I've left a few comments, but also it looks like the dates need tweaking in the new tests as they are not passing (see the [latest CI run](https://github.com/google/site-kit-wp/actions/runs/4006506114/jobs/6878114301)).

**Comment by techanvil** [comment]
I've made the last few changes myself as @asvinb has moved off the project.
